src/codegen/baseline_jit.h

```diff
@@ -0,0 +1,249 @@
+//      second:  return value only used if first == 0
+//
+// Basic layout of generated code block is:
+// entry_code:
+//      push   %rbp                 ; setup frame pointer
+//      mov    %rsp,%rbp
+//      sub    $0x108,%rsp          ; setup scratch, 0x108 = scratch_size + 8 (=stack alignment)
+//      push   %rdi                 ; save the pointer to ASTInterpreter instance
+//      jmpq   *0x8(%rsi)           ; jump to block->code
+//                                      possible values: first_JitFragment, second_JitFragment,...
+//
+// first_JitFragment:
+//      ...
+//    ; Side exit (e.g. if <val> <block1> else <block2>
+//      movabs $0x1270014108,%rcx   ; rcx = True
+//      cmp    %rax,%rcx            ; rax == True
+//      jne    end_side_exit
+//      mov    %rax,0x10(%rsp)      ;
+//      movabs $0x215bb60,%rax      ; rax = CFGBlock* to call next (rax is also the 1. return reg)
+//      mov    0x8(%rax),%rsi       ; load CFGBlock->code
+//      test   %rsi,%rsi            ; CFGBlock->code == 0
+//      je     epilog               ; exit to interpreter if code == 0
+//      jmpq   *0x8(%rax)           ; jump to new jit fragment (e.g second_JitFragment)
+//    end_side_exit:
+//      ....
+
+// second_JitFragment:
+//      ...
+//
+// nth_JitFragment:
+//      ...                         ; direct jump previous JITed block
+//      jmp first_JitFragment
+//
+// epilog:                          ; code which jumps to epilog has to make sure that
+//                                  ; rax contains the next block to execute
+//                                  : or 0 if we are finished but then rdx must contain the Box* value to return
+//      leave
+//      ret
+//
+class JitCodeBlock {
+private:
+    static constexpr int scratch_size = 256;
+    static constexpr int code_size = 4096 * 2;
+    static constexpr int epilog_size = 2; // size of [leave, ret] in bytes
+
+    void* code;
+    int entry_offset;
+    int epilog_offset;
+    assembler::Assembler a;
+    bool iscurrently_tracing;
+    bool asm_failed;
+
+public:
+    JitCodeBlock(llvm::StringRef name);
+
+    std::unique_ptr<JitFragment> newFragment(CFGBlock* block, int jump_offset = 0);
+    bool shouldCreateNewBlock() const { return asm_failed || a.bytesLeft() < 128; }
+    void fragmentFinished(int size, bool not_enough_space);
+};
+
+class JitFragment : public Rewriter {
+private:
+    CFGBlock* block;
+    int code_offset;
+    int epilog_offset;
+    int continue_jmp_offset;
+    void* entry_code;
+    JitCodeBlock& code_block;
+    RewriterVar* interp;
+    llvm::DenseMap<InternedString, RewriterVar*> local_syms;
+
+public:
+    JitFragment(CFGBlock* block, ICSlotRewrite* rewrite, int code_offset, int epilog_offset, void* entry_code,
+                JitCodeBlock& code_block);
+
+    RewriterVar* imm(uint64_t val);
+    RewriterVar* imm(void* val);
+
+
+    RewriterVar* emitAugbinop(Value lhs, Value rhs, int op_type);
```

Comment:
Super minor, but do we ever use the actual value part of the arguments, ie could we just pass the RewriterVar part of them?